### PR TITLE
Feature/174 add new population dataset

### PR DESCRIPTION
**Pull request by Tomkourou**
# Closes # (if applicable).

## Checklist

- [x] I tested my contribution locally and it seems to work fine.
- [x] Code and workflow changes are sufficiently documented.
- [x] Newly introduced dependencies are added to `envs/environment.yaml` and `envs/environment.docs.yaml`.
- [x] Changes in configuration options are added in all of `config.default.yaml`, `config.tutorial.yaml`, and `test/config.test1.yaml`.
- [ ] Changes in configuration options are also documented in `doc/configtables/*.csv` and line references are adjusted in `doc/configuration.rst` and `doc/tutorial.rst`.
- [x] A note for the release notes `doc/release_notes.rst` is amended in the format of previous release notes, including reference to the requested PR.

### CODE REVIEW

**Comment by pz-max** [comment]
Discussed suggestions from the weekly open workflow meeting:
- API version to extract the data is prefered
- We need to check if we can also extract low-resolution data 1x1km (or even 10x10km) data with the API
- If the 1x1km resolution is not possible, we need to keep both world_pop extraction methods 100mx100m and the 1x1km version.

FYI @davide-f

**Comment by davide-f** [comment]
@Tomkourou The contribution looks very nice and I look forward to the refinings, as discussed. Please, there are few merge conflicts that may be solved, luckily easy to address; they most likely are due to the use of the pre-commit.
Please, to do so, make sure that:
- you use the pypsa-africa environment
- if you add commits manually, make sure tu use the environment when adding commits [before doing `git commit -m "{message}"`. Do conda activate pypsa-africa)
- if you use VSCode to do so, make sure to setup the python interpreter "View->Command Pallette-> Python Interpreter -> Select pypsa-africa"
- if that may not be enough, after having enabled the pypsa-africa environment; activate the pre-commit using "pre-commit install"
```
conda activate pypsa-africa
pre-commit install
```

**Comment by Tomkourou** [comment]
Hi @davide-f , I have been experiencing problems with pre-commit so I just reset and pushed my changes without it in the meantime. I just reinstalled the environment so it should be the latest version. I installed pre-commit against and installed the hooks. Trying to commit changes throws up the following error:

```
Traceback (most recent call last):
  File "/home/thomas/.cache/pre-commit/repo1j9hz3vy/py_env-python3.10/bin/black", line 8, in <module>
    sys.exit(patched_main())
  File "/home/thomas/.cache/pre-commit/repo1j9hz3vy/py_env-python3.10/lib/python3.10/site-packages/black/__init__.py", line 1282, in patched_main
    patch_click()
  File "/home/thomas/.cache/pre-commit/repo1j9hz3vy/py_env-python3.10/lib/python3.10/site-packages/black/__init__.py", line 1268, in patch_click
    from click import _unicodefun  # type: ignore
ImportError: cannot import name '_unicodefun' from 'click' (/home/thomas/.cache/pre-commit/repo1j9hz3vy/py_env-python3.10/lib/python3.10/site-packages/click/__init__.py)
```

**Comment by davide-f** [comment]
> Hi @davide-f , I have been experiencing problems with pre-commit so I just reset and pushed my changes without it in the meantime. I just reinstalled the environment so it should be the latest version. I installed pre-commit against and installed the hooks. Trying to commit changes throws up the following error:
> 
> ```
> Traceback (most recent call last):
>   File "/home/thomas/.cache/pre-commit/repo1j9hz3vy/py_env-python3.10/bin/black", line 8, in <module>
>     sys.exit(patched_main())
>   File "/home/thomas/.cache/pre-commit/repo1j9hz3vy/py_env-python3.10/lib/python3.10/site-packages/black/__init__.py", line 1282, in patched_main
>     patch_click()
>   File "/home/thomas/.cache/pre-commit/repo1j9hz3vy/py_env-python3.10/lib/python3.10/site-packages/black/__init__.py", line 1268, in patch_click
>     from click import _unicodefun  # type: ignore
> ImportError: cannot import name '_unicodefun' from 'click' (/home/thomas/.cache/pre-commit/repo1j9hz3vy/py_env-python3.10/lib/python3.10/site-packages/click/__init__.py)
> ```

That's new! We experienced something like that related to compatibility with black; we solved the issue by changing the black version in the [pre-commit settings](https://github.com/pypsa-meets-africa/pypsa-africa/blob/main/.pre-commit-config.yaml).
Could you post the conda list?

**Comment by Tomkourou** [comment]
> > Hi @davide-f , I have been experiencing problems with pre-commit so I just reset and pushed my changes without it in the meantime. I just reinstalled the environment so it should be the latest version. I installed pre-commit against and installed the hooks. Trying to commit changes throws up the following error:
> > ```
> > Traceback (most recent call last):
> >   File "/home/thomas/.cache/pre-commit/repo1j9hz3vy/py_env-python3.10/bin/black", line 8, in <module>
> >     sys.exit(patched_main())
> >   File "/home/thomas/.cache/pre-commit/repo1j9hz3vy/py_env-python3.10/lib/python3.10/site-packages/black/__init__.py", line 1282, in patched_main
> >     patch_click()
> >   File "/home/thomas/.cache/pre-commit/repo1j9hz3vy/py_env-python3.10/lib/python3.10/site-packages/black/__init__.py", line 1268, in patch_click
> >     from click import _unicodefun  # type: ignore
> > ImportError: cannot import name '_unicodefun' from 'click' (/home/thomas/.cache/pre-commit/repo1j9hz3vy/py_env-python3.10/lib/python3.10/site-packages/click/__init__.py)
> > ```
> 
> That's new! We experienced something like that related to compatibility with black; we solved the issue by changing the black version in the [pre-commit settings](https://github.com/pypsa-meets-africa/pypsa-africa/blob/main/.pre-commit-config.yaml). Could you post the conda list?

Changing black version fixed the issue thanks!

File: scripts/build_shapes.py [hunk 1]
```diff
@@ -302,52 +302,86 @@ def eez(countries, country_shapes, EEZ_gpkg, out_logging=False, distance=0.01):
     return ret_df
 
 
-def download_WorldPop(
-    country_code, year=2020, update=False, out_logging=False, size_min=300
-):
-    """
-    Download tiff file for each country code
+if WorldPop_method == False:
```
Comment (davide-f) [inline scripts/build_shapes.py]: I'd recommend to create two functions:

def download_WorldPop_default(...) : function that would be basically the old one
def download_WorldPop_API(...) : function that would be basically the new one that you propose

The function download_WorldPop, instead may be modified to be something like:

``` python
def download_WorldPop(
        WorldPop_method,
        country_code,
        year=2020,
        update=False,
        out_logging=False,
        size_min=300,
    ):
....
    if WorldPop_method="API":
        download_WorldPop_API(...)
    else:
        download_WorldPop_standard(...)
```
With respect to the current implementation, wherever download_WorldPop is called, there is the need to specify the method, which can be loaded from the configuration file in the main section (at the end of the file)

What do you think? Feel free to comment and or add suggestions
Thank you for your contribution!

File: test/config.test1.yaml [hunk 2]
```diff
@@ -50,19 +51,23 @@ build_shape_options:
   out_logging: true    # When true, logging is printed to console
   year: 2020    # reference year used to derive shapes, info on population and info on GDP
   nprocesses: 5    # number of processes to be used in build_shapes
+  worldpop_api: false
```
Comment (davide-f) [inline test/config.test1.yaml]: may we call if world_pop_method and having as values "api" or "default"?
What do you think?

**Comment by davide-f** [comment]
@Tomkourou The CI is failing because of the definition of worldpop method that is not inside a function that is called.
I added a comment, if it is not clear, please, I'm available also on discord for chatting :)

File: scripts/build_shapes.py [hunk 3]
```diff
@@ -707,6 +800,7 @@ def gadm(
     year = snakemake.config["build_shape_options"]["year"]
     nprocesses = snakemake.config["build_shape_options"]["nprocesses"]
     EEZ_gpkg = snakemake.input["eez"]
+    worldpop_method = snakemake.config["build_shape_options"]["worldpop_api"]
```
Comment (pz-max) [inline scripts/build_shapes.py]: The config read needs to be `snakemake.config["build_shape_options"]["worldpop_method"]`

File: scripts/build_shapes.py [hunk 4]
```diff
@@ -303,7 +303,47 @@ def eez(countries, country_shapes, EEZ_gpkg, out_logging=False, distance=0.01):
...
+            Files downloaded from https://data.worldpop.org/ datasets WorldPop UN adjusted
+        year : int
+            Year of the data to download
+        update : bool
+            Update = true, forces re-download of files
+        size_min : int
+            Minimum size of each file to download
+
+    """
+    if worldpop_method == "API":
```
Comment (pz-max) [inline scripts/build_shapes.py]: maybe use normal letters instead of capital ones?

File: scripts/build_shapes.py [hunk 5]
```diff
@@ -303,7 +303,47 @@ def eez(countries, country_shapes, EEZ_gpkg, out_logging=False, distance=0.01):
...
 def download_WorldPop(
-    country_code, year=2020, update=False, out_logging=False, size_min=300
+    worldpop_method,
+    country_code,
+    year=2020,
+    update=False,
+    out_logging=False,
+    size_min=300,
+):
+
```
Comment (pz-max) [inline scripts/build_shapes.py]: You could follow the docstring convention. Some empty spaces should be removed. See:
https://peps.python.org/pep-0257/

File: scripts/build_shapes.py [hunk 6]
```diff
@@ -303,7 +303,47 @@ def eez(countries, country_shapes, EEZ_gpkg, out_logging=False, distance=0.01):
...
+        country_code : str
+            Two letter country codes of the downloaded files.
+            Files downloaded from https://data.worldpop.org/ datasets WorldPop UN adjusted
+        year : int
+            Year of the data to download
+        update : bool
+            Update = true, forces re-download of files
+        size_min : int
+            Minimum size of each file to download
+
```
Comment (pz-max) [inline scripts/build_shapes.py]: empty line

File: scripts/build_shapes.py [hunk 7]
```diff
@@ -303,7 +303,47 @@ def eez(countries, country_shapes, EEZ_gpkg, out_logging=False, distance=0.01):
...
+    """
+
+    Download Worldpop using either the standard method or the API method.
+
+        Parameters
+        ----------
+        worldpop_method: str
+             worldpop_method = "API" will use the API method to access the WorldPop 100mx100m dataset.  worldpop_method = "standard" will use the standard method to access the WorldPop 1KMx1KM dataset.
+
```
Comment (pz-max) [inline scripts/build_shapes.py]: empty line

File: scripts/build_shapes.py [hunk 8]
```diff
@@ -320,6 +360,9 @@ def download_WorldPop(
     size_min : int
         Minimum size of each file to download
 
+    api_method: bool
```
Comment (pz-max) [inline scripts/build_shapes.py]: I think this is an artifact?

**Review by pz-max — COMMENTED** [review]
Great job. Almost there :+1:  Some comments below. 
There is also a HiGHS submodule. Would suggest to remove it from this PR.

**Comment by pz-max** [comment]
The issue is here (missing the required read in for the function):
https://github.com/pypsa-meets-africa/pypsa-africa/blob/3d88d8ca2153b5a06f053984a7833afc179096af/scripts/build_shapes.py#L643-L645

Should we have a short session? Happy to check out your coding setup and guide you to get the PR done. Maybe we can even record the session & upload it on Youtube (helps also other users :) )

File: scripts/build_shapes.py [hunk 9]
```diff
@@ -303,7 +303,43 @@ def eez(countries, country_shapes, EEZ_gpkg, out_logging=False, distance=0.01):
...
+            Files downloaded from https://data.worldpop.org/ datasets WorldPop UN adjusted
+        year : int
+            Year of the data to download
+        update : bool
+            Update = true, forces re-download of files
+        size_min : int
+            Minimum size of each file to download
+    """
+    if worldpop_method == "api":
+        download_WorldPop_API(country_code, year, update, out_logging, size_min)
```
Comment (davide-f) [inline scripts/build_shapes.py]: @Tomkourou I think the CI error is due to the missing return statement here (line 331) and at line 334.
If you add them, the CI should move on

**Review by davide-f — APPROVED** [review]

**Comment by pz-max** [comment]
Welcome @Tomkourou to the contributor team :100:

**Comment by Tomkourou** [comment]
Thanks for all the help with this PR guys!

**Pull request merged**